### PR TITLE
Rangeroaringset optimizations

### PR DESCRIPTION
**Pull request by aliszka**
### What's being changed:
Change the way segments of roaringsetrange strategy are filtered:
Now each segment is filtered individually and concurrently providing its own filtering subresults. At the end subresults are merge together in order (oldest segment to newest), taking into account deletions occurred to calculate one final result.

Sroar bitmap library is updated to provide faster and less memory demanding methods to merge bitmaps.

Provides 2 implementations of segment cursors: mmap + pread.

### Review checklist

- [ ] Documentation has been updated, if necessary. Link to changed documentation:
- [ ] Chaos pipeline run or not necessary. Link to pipeline:
- [x] All new code is covered by tests where it is reasonable.
- [x] Performance tests have been run or not necessary.

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -70,7 +70,7 @@ require (
 	github.com/vmihailenco/msgpack/v5 v5.4.1
 	github.com/weaviate/fgprof v0.0.0-20240724124431-aab99fd9c876
 	github.com/weaviate/s5cmd/v2 v2.0.1
-	github.com/weaviate/sroar v0.0.2
+	github.com/weaviate/sroar v0.0.3-0.20240820105622-36ab72883cb3
```
Comment (antas-marcin) [inline go.mod:73]: I think that here we can switch to recently released `0.0.3` version

File: adapters/repos/db/roaringsetrange/segment_reader.go [hunk 2]
```diff
@@ -23,14 +23,23 @@ import (
...
 }
 
 func NewSegmentReader(cursor *GaplessSegmentCursor) *SegmentReader {
+	return NewSegmentReaderConcurrent(cursor, 1)
+}
+
+func NewSegmentReaderConcurrent(cursor *GaplessSegmentCursor, concurrency int) *SegmentReader {
+	sroarBufs := make([][]uint16, concurrency)
+	for i := range sroarBufs {
+		sroarBufs[i] = make([]uint16, 4100) // sroar.maxContainerSize
```
Comment (antas-marcin) [inline adapters/repos/db/roaringsetrange/segment_reader.go:37]: 4100 == sroar.maxContainerSize, yes?

Comment (aliszka) [reply adapters/repos/db/roaringsetrange/segment_reader.go:37]: correct. hardcoded, cause `maxContainerSize` is not exported

**Review by antas-marcin — DISMISSED** [review]

File: adapters/repos/db/roaringsetrange/memtable_reader.go [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+
+func (r *MemtableReader) read(predicate func(k uint64) bool) roaringset.BitmapLayer {
+	additions := sroar.NewBitmap()
+	deletions := sroar.NewBitmap()
+
+	for v, k := range r.memtable.additions {
+		if predicate(k) {
+			additions.Set(v)
+		}
+		deletions.Set(v)
```
Comment (jeroiraz) [inline adapters/repos/db/roaringsetrange/memtable_reader.go:70]: `else` missing?

Comment (aliszka) [reply adapters/repos/db/roaringsetrange/memtable_reader.go:70]: nope. deletions are always added

File: adapters/repos/db/roaringsetrange/segment_cursor.go [hunk 4]
```diff
@@ -53,3 +63,153 @@ func (c *SegmentCursor) Next() (uint8, roaringset.BitmapLayer, bool) {
...
+	return c.read(false)
+}
+
+func (c *SegmentCursorPread) read(isFirst bool) (uint8, roaringset.BitmapLayer, bool) {
+	_, err := io.ReadFull(c.reader, c.lenBuf)
+	if err == io.EOF {
+		return 0, roaringset.BitmapLayer{}, false
+	}
+	if err != nil {
+		panic(fmt.Errorf("SegmentCursorReader::Next reading node length: %w", err))
```
Comment (jeroiraz) [inline adapters/repos/db/roaringsetrange/segment_cursor.go:120]: erroring would be preferred over panicking whenever possible

File: adapters/repos/db/roaringsetrange/segment_cursor.go [hunk 5]
```diff
@@ -53,3 +63,153 @@ func (c *SegmentCursor) Next() (uint8, roaringset.BitmapLayer, bool) {
...
+	}
+
+	nodeLen := binary.LittleEndian.Uint64(c.lenBuf)
+	nodeBuf := c.getNodeBuf(int(nodeLen))
+
+	_, err = io.ReadFull(c.reader, nodeBuf[8:])
+	if err != nil {
+		panic(fmt.Errorf("SegmentCursorReader::Next reading node: %w", err))
```
Comment (jeroiraz) [inline adapters/repos/db/roaringsetrange/segment_cursor.go:128]: erroring would be preferred over panicking whenever possible

File: adapters/repos/db/lsmkv/bucket_roaring_set_range.go [hunk 6]
```diff
@@ -32,3 +43,33 @@ func (b *Bucket) RoaringSetRangeRemove(key uint64, values ...uint64) error {
 
 	return b.active.roaringSetRangeRemove(key, values...)
 }
+
+type ReaderRoaringSetRange interface {
+	Read(ctx context.Context, value uint64, operator filters.Operator) (*sroar.Bitmap, error)
+	Close()
+}
+
+func (b *Bucket) ReaderRoaringSetRange() ReaderRoaringSetRange {
```
Comment (jeroiraz) [inline adapters/repos/db/lsmkv/bucket_roaring_set_range.go:52]: I'd think the active memtable is not included so to not block insertions, isn't it?

File: adapters/repos/db/lsmkv/bucket_roaring_set_range.go [hunk 7]
```diff
@@ -32,3 +43,33 @@ func (b *Bucket) RoaringSetRangeRemove(key uint64, values ...uint64) error {
...
+
+	// we have a flush-RLock, so we have the guarantee that the flushing state
+	// will not change for the lifetime of the cursor, thus there can only be two
+	// states: either a flushing memtable currently exists - or it doesn't
+	if b.flushing != nil {
+		readers = append(readers, b.flushing.newRoaringSetRangeReader())
+	}
+	readers = append(readers, b.active.newRoaringSetRangeReader())
+
+	concurrency := 4
```
Comment (jeroiraz) [inline adapters/repos/db/lsmkv/bucket_roaring_set_range.go:67]: just wondering the reason of the concurrency level

File: adapters/repos/db/roaringsetrange/segment_cursor.go (hunk 4 above)
Comment (aliszka) [reply adapters/repos/db/roaringsetrange/segment_cursor.go:120]: Agreed. Even had errors in cursor's method signatures, but api and error handling got so messy, I've removed it in favour of completing implementation. Can try again with errors in separate PR

File: adapters/repos/db/roaringsetrange/segment_cursor.go (hunk 5 above)
Comment (aliszka) [reply adapters/repos/db/roaringsetrange/segment_cursor.go:128]: Agreed. Even had errors in cursor's method signatures, but api and error handling got so messy, I've removed it in favour of completing implementation. Can try again with errors in separate PR

File: adapters/repos/db/roaringsetrange/memtable_reader.go (hunk 3 above)
Comment (aliszka) [reply adapters/repos/db/roaringsetrange/memtable_reader.go:70]: Newer values replace older ones. Deletions tell which values added in the past should be ignored, therefore except extending deletions explicitly on delete, they are extended implicitly on add.

File: adapters/repos/db/lsmkv/bucket_roaring_set_range.go (hunk 7 above)
Comment (aliszka) [reply adapters/repos/db/lsmkv/bucket_roaring_set_range.go:67]: that setting indeed could be changed (or explained). will change that to numcpu/2.
with conc=4 I got the best results locally. further increasing concurrency level improved performance only slightly or not at all.

File: adapters/repos/db/lsmkv/bucket_roaring_set_range.go (hunk 6 above)
Comment (aliszka) [reply adapters/repos/db/lsmkv/bucket_roaring_set_range.go:52]: active memtable is added in line 65. flushing memtable is added in line 63 (if present).
or you have something else in mind?

Comment (jeroiraz) [reply adapters/repos/db/lsmkv/bucket_roaring_set_range.go:50]: ah, haven't seen that line when active memtable is added

**Review by jeroiraz — APPROVED** [review]
LGTM!

**Pull request merged**